### PR TITLE
Replace LRU cache blacklist implementation with a time cache

### PR DESCRIPTION
**Pull request by aarshkshah1992**
@vyzo

### CODE REVIEW

File: blacklist.go [hunk 1]
```diff
@@ -28,26 +30,21 @@ func (b MapBlacklist) Contains(p peer.ID) bool {
 	return ok
 }
 
-// LRUBlacklist is a blacklist implementation using an LRU cache
-type LRUBlacklist struct {
-	lru *lru.Cache
+// TimeCachedBlacklist is a blacklist implementation using a time cache
+type TimeCachedBlacklist struct {
+	tc *timecache.TimeCache
```
Comment (vyzo) [inline blacklist.go:37]: it is not thread-safe, we need a mutex.

File: blacklist.go [hunk 2]
```diff
@@ -28,26 +30,21 @@ func (b MapBlacklist) Contains(p peer.ID) bool {
...
+// NewTimeCachedBlacklist creates a new TimeCachedBlacklist with the given expiry duration
+func NewTimeCachedBlacklist(expiry time.Duration) (Blacklist, error) {
+	b := &TimeCachedBlacklist{timecache.NewTimeCache(expiry)}
 	return b, nil
 }
 
-func (b LRUBlacklist) Add(p peer.ID) {
-	b.lru.Add(p, nil)
+func (b TimeCachedBlacklist) Add(p peer.ID) {
+	b.tc.Add(p.String())
```
Comment (vyzo) [inline blacklist.go:50]: mutex lock here

File: blacklist.go [hunk 3]
```diff
@@ -28,26 +30,21 @@ func (b MapBlacklist) Contains(p peer.ID) bool {
...
 }
 
-func (b LRUBlacklist) Contains(p peer.ID) bool {
-	return b.lru.Contains(p)
+func (b TimeCachedBlacklist) Contains(p peer.ID) bool {
+	return b.tc.Has(p.String())
```
Comment (vyzo) [inline blacklist.go:57]: mutex lock here too.

**Review by vyzo — CHANGES_REQUESTED** [review]
this needs a mutex, timecache is not thread-safe.

**Comment by aarshkshah1992** [comment]
@vyzo  Have added a mutex. Thanks.

File: blacklist.go [hunk 4]
```diff
@@ -32,19 +33,25 @@ func (b MapBlacklist) Contains(p peer.ID) bool {
...
-	b := &TimeCachedBlacklist{timecache.NewTimeCache(expiry)}
+	b := &TimeCachedBlacklist{tc: timecache.NewTimeCache(expiry)}
 	return b, nil
 }
 
-func (b TimeCachedBlacklist) Add(p peer.ID) {
+func (b *TimeCachedBlacklist) Add(p peer.ID) {
+	b.Lock()
 	b.tc.Add(p.String())
+	b.Unlock()
```
Comment (vyzo) [inline blacklist.go]: defer this, it's more idiomatic.

Comment (aarshkshah1992) [reply blacklist.go]: Done.

**Review by vyzo — APPROVED** [review]

**Comment by vyzo** [comment]
thank you, will merge as soon as travis finishes.

**Pull request merged**